### PR TITLE
chore(fusdc): settle `ForwardFailed`, minted while `Advancing`, and minted early txs

### LINKED ISSUE #10625: fusdc: `ForwardFailed` state
## What is the Problem Being Solved?

In FUSDC, the contact does not distinguish between failed forwards and successful ones.

## Description of the Design

- do not call `.forwarded ()` on initiation of forward
   - _we considered a `Forwarding` state, but the extra vstorage write seems unnecessary_
- transition to `Forwarded` **_after_** the transfer vow has fulfilled
- transition to `ForwardFailed` after transfer vow has rejected

### PR DESCRIPTION
**Pull request by 0xpatrickdev**
closes: #10625

## Description
- fix: ensure "minted while `Advancing`" results in `Disburse` or `Forward`
- feat: handle "unknown mint" by forwarding if and when Evidence is reported
   - via `Settler.forwardIfMinted()` called by `Advancer`
- chore: capture `ForwardFailed` (terminal) state if `Forward` fails

### Security Considerations
None new introduced, these behaviors are consistent with the product spec.

### Scaling Considerations
The `mintedEarly` mapStore could grow large if an attacker spams the `settlementAccount` with uusdc

### Documentation Considerations
None

### Testing Considerations
Includes tests for all new behaviors.

### Upgrade Considerations
Targeting FUSDC's first release

### CODE REVIEW

File: packages/fast-usdc/src/exos/advancer.js [hunk 1]
```diff
@@ -155,6 +153,24 @@ export const prepareAdvancerKit = (
...
+            } = evidence;
+            try {
+              notifyFacet.checkMintedEarly(
+                destination,
+                forwardingAddress,
+                fullAmount,
+                txHash,
+              );
+            } catch (e) {
+              // settlement already received; tx will Forward.
```
Comment (dckc) [inline packages/fast-usdc/src/exos/advancer.js]: assuming all exceptions have this meaning seems inconsistent with https://github.com/Agoric/agoric-sdk/wiki/Errors-and-Control-Flow

File: packages/fast-usdc/src/exos/settler.js [hunk 2]
```diff
@@ -227,6 +237,23 @@ export const prepareSettler = (
...
+         * @throws {Error} if minted early, so advancer doesn't advance
+         */
+        checkMintedEarly(destination, forwardingAddress, fullAmount, txHash) {
+          const { value: fullValue } = fullAmount;
+          const key = makeMintedEarlyKey(forwardingAddress, fullValue);
+          const { mintedEarly } = this.state;
+          if (mintedEarly.has(key)) {
+            mintedEarly.delete(key);
+            void this.facets.self.forward(txHash, fullValue, destination.value);
+            throw Fail`${q(txHash)} minted before Observed`;
```
Comment (dckc) [inline packages/fast-usdc/src/exos/settler.js]: This should be "raising a non-error" per [house style on Errors and Control Flow](https://github.com/Agoric/agoric-sdk/wiki/Errors-and-Control-Flow)

**Review by dckc — COMMENTED** [review]
a couple notes before our chat

File: packages/fast-usdc/src/exos/settler.js [hunk 3]
```diff
@@ -174,19 +179,23 @@ export const prepareSettler = (
...
-              return self.forward(found.txHash, nfa, amount, EUD);
+              return self.forward(found.txHash, amount, EUD);
 
             case undefined:
             default:
-              log('⚠️ tap: no status for ', nfa, amount);
+              log('⚠️ tap: minted before observed', nfa, amount);
+              // XXX consider capturing in vstorage
+              // we would need a new key, as this does not have a txHash
+              this.state.mintedEarly.add(makeMintedEarlyKey(nfa, amount));
```
Comment (0xpatrickdev) [inline packages/fast-usdc/src/exos/settler.js:204]: I realized while writing this PR desc:

> The mintedEarly mapStore could grow large if an attacker spams the settlementAccount with uusdc

Worth adding a check here to see if the advance exceeds min fees? That should like help with 1uusdc spam.

File: packages/fast-usdc/src/exos/settler.js [hunk 4]
```diff
@@ -206,19 +215,19 @@ export const prepareSettler = (
         ) {
           const { mintedEarly } = this.state;
           const { value: fullValue } = fullAmount;
+          // XXX i think this only contains Advancing txs - should this be a separate store?
```
Comment (0xpatrickdev) [inline packages/fast-usdc/src/exos/settler.js]: comment for reviewers, not something to check in

File: packages/fast-usdc/test/fast-usdc.contract.test.ts [hunk 5]
```diff
@@ -787,29 +811,122 @@ test.serial('Settlement for unknown transaction (operator down)', async t => {
...
     ],
     '20 USDC arrive at the settlement account',
   );
-  t.deepEqual(bridgeTraffic.local, [], 'no IBC transfers');
 
   await transmitTransferAck();
+  t.deepEqual(bridgeTraffic.local, [], 'no IBC transfers');
+
+  // activate oracles and submit evidence; expect Settler to forward (slow path)
+  // 'C12 - Contract MUST only pay back the Pool (fees) only if they started the advance before USDC is minted',
```
Comment (0xpatrickdev) [inline packages/fast-usdc/test/fast-usdc.contract.test.ts:876]: This is maybe not _entirely_ relevant. We could adjust the test to fail (`AdvanceFailed`), and we would not pay fees to the pool here. This is covered in `settler.test.ts`, though.

File: packages/fast-usdc/src/exos/settler.js [hunk 6]
```diff
@@ -279,35 +265,23 @@ export const prepareSettler = (
...
          * @param {unknown} reason
-         * @param {SettlerTransferCtx} ctx
+         * @param {EvmHash} txHash
          */
-        onRejected(reason, ctx) {
-          log('⚠️ transfer rejected!', reason, ctx);
-          // const { txHash, nfa, amount } = ctx;
+        onRejected(reason, txHash) {
+          log('⚠️ transfer rejected!', reason, txHash);
           // TODO(#10510): statusManager.forwardFailed(txHash, nfa, amount);
```
Comment (dckc) [inline packages/fast-usdc/src/exos/settler.js]: #10510 is closed. should this TODO still be here?

File: packages/fast-usdc/src/exos/status-manager.js [hunk 7]
```diff
@@ -69,6 +69,7 @@ export const prepareStatusManager = (
   txnsNode,
   {
     marshaller,
+    // eslint-disable-next-line no-unused-vars
     log = makeTracer('Advancer', true),
```
Comment (dckc) [inline packages/fast-usdc/src/exos/status-manager.js]: this tracer shouldn't call itself `'Advancer'`.
out of scope of this PR, but since you're in the neighborhood...

File: packages/fast-usdc/src/exos/status-manager.js [hunk 8]
```diff
@@ -69,6 +69,7 @@ export const prepareStatusManager = (
   txnsNode,
   {
     marshaller,
+    // eslint-disable-next-line no-unused-vars
```
Comment (dckc) [inline packages/fast-usdc/src/exos/status-manager.js:72]: we should use the log/tracer, shouldn't we?

File: packages/fast-usdc/src/exos/status-manager.js [hunk 9]
```diff
@@ -325,19 +324,10 @@ export const prepareStatusManager = (
       /**
        * Mark a transaction as `FORWARDED`
        *
-       * @param {EvmHash | undefined} txHash - undefined in case mint before observed
-       * @param {NobleAddress} nfa
-       * @param {bigint} amount
+       * @param {EvmHash} txHash - undefined in case mint before observed
```
Comment (dckc) [inline packages/fast-usdc/src/exos/status-manager.js]: the "undefined in case..." docstring looks outdated

Comment (dckc) [reply packages/fast-usdc/src/exos/status-manager.js]: I suppose the preconditions for `forwarded(...)` are in a state transition diagram somewhere, but it would be nice to have them here locally.

In particular, this would throw if you call it twice on the same txHash, right?

File: packages/fast-usdc/src/exos/settler.js [hunk 10]
```diff
@@ -199,12 +209,17 @@ export const prepareSettler = (
         ) {
           const { mintedEarly } = this.state;
           const { value: fullValue } = fullAmount;
+          // XXX i think this only contains Advancing txs - should this be a separate store?
           const key = makeMintedEarlyKey(forwardingAddress, fullValue);
           if (mintedEarly.has(key)) {
             mintedEarly.delete(key);
             if (success) {
+              // TODO: does not write `ADVANCED` to vstorage
```
Comment (dckc) [inline packages/fast-usdc/src/exos/settler.js]: TODO when? in this PR?

File: packages/fast-usdc/src/exos/status-manager.js [hunk 11]
```diff
@@ -202,11 +202,7 @@ export const prepareStatusManager = (
         M.or(
           {
             txHash: EvmHashShape,
-            status: M.or(
-              PendingTxStatus.Advanced,
-              PendingTxStatus.AdvanceFailed,
-              PendingTxStatus.Observed,
-            ),
+            status: M.or(...Object.values(PendingTxStatus)),
```
Comment (dckc) [inline packages/fast-usdc/src/exos/status-manager.js]: nice.

File: packages/fast-usdc/src/exos/status-manager.js [hunk 12]
```diff
@@ -277,7 +273,10 @@ export const prepareStatusManager = (
       },
 
       /**
-       * Remove and return an `ADVANCED` or `OBSERVED` tx waiting to be `SETTLED`.
+       * Remove and return the _first_ `PendingSettleTx` that matches the observed
+       * settlement. This is not necessarily an exact match with the original
```
Comment (dckc) [inline packages/fast-usdc/src/exos/status-manager.js]: what does "not necessarily an exact match" mean? isn't "same forwarding account and amount" the relevant definition of "matching"?

File: packages/fast-usdc/test/fast-usdc.contract.test.ts [hunk 13]
```diff
@@ -787,29 +811,122 @@ test.serial('Settlement for unknown transaction (operator down)', async t => {
...
+  t.is(
+    outgoingForwardMessage.token.amount,
+    String(sent.tx.amount),
+    'full amount is transferred via `.forward()`',
+  );
+
+  const forwardInfo = JSON.parse(outgoingForwardMessage.memo).forward;
+  t.is(forwardInfo.receiver, EUD, 'receiver is osmo10tt0');
+
+  // in lieu of transmitTransferAck so we can set a nonce that matches our initial Advance
```
Comment (0xpatrickdev) [inline packages/fast-usdc/test/fast-usdc.contract.test.ts:906]: considered parameterizing like: `transmitTransferAck(-1)` but came up short. I think it's only the sequence that difference, not the index of `localBridgeMessage`.

File: packages/fast-usdc/test/exos/settler.test.ts [hunk 14]
```diff
@@ -493,3 +493,69 @@ test('Settlement for Advancing transaction (advance fails)', async t => {
...
+  const mockE = Error('no connection info found');
+  accounts.settlement.transferVResolver.reject(mockE);
+  await eventLoopIteration();
+  t.deepEqual(storage.getDeserialized(`fun.txns.${cctpTxEvidence.txHash}`), [
+    { evidence: cctpTxEvidence, status: 'OBSERVED' },
+    { status: 'FORWARD_FAILED' },
+  ]);
+});
+
+test.todo('create facet methods');
```
Comment (dckc) [inline packages/fast-usdc/test/exos/settler.test.ts]: creator facet?

**Review by dckc — COMMENTED** [review]
here are some comments based on one read / skim

File: packages/fast-usdc/src/exos/settler.js [hunk 15]
```diff
@@ -227,14 +236,38 @@ export const prepareSettler = (
...
+          const key = makeMintedEarlyKey(forwardingAddress, fullValue);
+          const { mintedEarly } = this.state;
+          if (mintedEarly.has(key)) {
+            log(
+              'matched minted early key, initiating forward',
+              forwardingAddress,
+              fullValue,
+            );
+            mintedEarly.delete(key);
+            // TODO: does not write `OBSERVED` to vstorage
```
Comment (0xpatrickdev) [inline packages/fast-usdc/src/exos/settler.js]: This also should make an entry in `seenTxs`. We can tackle this in the follow up PR that addresses vstorage TODOs

File: packages/fast-usdc/src/exos/status-manager.js (hunk 12 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/status-manager.js]: The ordering could be different if Alice requests two advances for the same amount.

File: packages/fast-usdc/src/exos/settler.js (hunk 10 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/settler.js]: #10750 but maybe will collapse into this

File: packages/fast-usdc/src/exos/status-manager.js (hunk 8 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/status-manager.js:72]: There's nothing currently logging anything in `statusManager`. vstorage seems to be the main way of communicating

File: packages/fast-usdc/src/exos/settler.js (hunk 6 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/settler.js]: I think this was removed in a later commit but will check again

File: packages/fast-usdc/src/exos/advancer.js [hunk 16]
```diff
@@ -161,6 +158,26 @@ export const prepareAdvancerKit = (
             const destination = chainHub.makeChainAddress(EUD);
 
             const fullAmount = toAmount(evidence.tx.amount);
+            const {
+              tx: { forwardingAddress },
+              txHash,
+            } = evidence;
+
+            const { borrowerFacet, notifyFacet, poolAccount } = this.state;
```
Comment (turadg) [inline packages/fast-usdc/src/exos/advancer.js:167]: sometime before this PR, but I'm surprised to see `Facet` in the name of the object. the consumer needn't be concerned with the structure in which the object resides.

I've added this to revisit in https://github.com/Agoric/agoric-sdk/issues/10432

File: packages/fast-usdc/src/exos/advancer.js [hunk 17]
```diff
@@ -161,6 +158,26 @@ export const prepareAdvancerKit = (
...
+            if (
+              notifyFacet.forwardIfMinted(
+                destination,
+                forwardingAddress,
+                fullAmount,
+                txHash,
+              )
+            ) {
```
Comment (turadg) [inline packages/fast-usdc/src/exos/advancer.js]: surprising to see the condition be the result of an action. please make more clear.
```suggestion
            const forwarded = notifyFacet.forwardIfMinted(
                destination,
                forwardingAddress,
                fullAmount,
                txHash,
              );
            if (forwarded)) {
```

File: packages/fast-usdc/src/exos/status-manager.js (hunk 12 above)
Comment (turadg) [reply packages/fast-usdc/src/exos/status-manager.js]: It is an exact match based on available data, which treats (EUD,Amount) interchangeably. 

I think think what you're getting at is that a mint deposit may be for any of the matching transactions and we can't distinguish between them, so we go with the first.

File: packages/fast-usdc/src/exos/advancer.js [hunk 18]
```diff
@@ -161,6 +158,26 @@ export const prepareAdvancerKit = (
...
+              // settlement already received; tx will Forward.
```
Comment (turadg) [inline packages/fast-usdc/src/exos/advancer.js]: It's not "will Forward" because that has started at this point. I don't think this comment adds enough over the log message and the `if (forwarded)` branch
```suggestion
```

File: packages/fast-usdc/src/exos/advancer.js [hunk 19]
```diff
@@ -202,7 +219,13 @@ export const prepareAdvancerKit = (
          */
         onFulfilled(result, ctx) {
           const { poolAccount, intermediateRecipient } = this.state;
-          const { destination, advanceAmount, ...detail } = ctx;
+          const {
+            destination,
+            advanceAmount,
+            // eslint-disable-next-line no-unused-vars
+            tmpSeat,
```
Comment (turadg) [inline packages/fast-usdc/src/exos/advancer.js]: consider,
```suggestion
            tmpSeat: _,
```

File: packages/fast-usdc/src/exos/settler.js [hunk 20]
```diff
@@ -69,24 +78,20 @@ export const prepareSettler = (
       }),
       notify: M.interface('SettlerNotifyI', {
         notifyAdvancingResult: M.call(
-          M.record(), // XXX fill in details TODO
+          makeAdvanceDetailsShape(USDC),
```
Comment (turadg) [inline packages/fast-usdc/src/exos/settler.js:85]: don't make the shape twice. Make it once in the top of `prepareSettler`.
```suggestion
          advanceDetailsShape,
```

File: packages/fast-usdc/src/exos/settler.js (hunk 4 above)
Comment (turadg) [reply packages/fast-usdc/src/exos/settler.js]: what's the alternative? pros/cons?

File: packages/fast-usdc/src/exos/settler.js [hunk 21]
```diff
@@ -206,19 +215,19 @@ export const prepareSettler = (
...
-              void this.facets.self.disburse(
-                txHash,
-                forwardingAddress,
-                fullValue,
-              );
+              // TODO: does not write `ADVANCED` to vstorage
+              // this is the "slow" experience, but we've already earmarked fees so disburse
+              void this.facets.self.disburse(txHash, fullValue);
             } else {
+              // TODO: does not write `ADVANCE_FAILED` to vstorage
```
Comment (turadg) [inline packages/fast-usdc/src/exos/settler.js]:
```suggestion
              // FIXME: does not write `ADVANCE_FAILED` to vstorage
```

File: packages/fast-usdc/src/exos/advancer.js (hunk 17 above)
Comment (turadg) [reply packages/fast-usdc/src/exos/advancer.js]: This name was weird to me and as I get through more of the review, I understand better why. The `notify` facet shouldn't be telling the settler what to do.

What is it notifying it of? It's telling the settler about a transaction and it needs to know what decision the settler has to say about it. So I think this should be,
```suggestion
            const forwarded = notifyFacet.checkForwarded(
                destination,
                forwardingAddress,
                fullAmount,
                txHash,
              );
            if (forwarded) {
              return;
             }
```

I took out the logging too because the settler already does that. I'm not as confident in that change.

File: packages/fast-usdc/src/exos/status-manager.js [hunk 22]
```diff
@@ -323,21 +320,18 @@ export const prepareStatusManager = (
...
-      forwarded(txHash, nfa, amount) {
-        if (txHash) {
-          void publishTxnRecord(txHash, harden({ status: TxStatus.Forwarded }));
-        } else {
-          // TODO store (early) `Minted` transactions to check against incoming evidence
-          log(
-            `⚠️ Forwarded minted amount ${amount} from account ${nfa} before it was observed.`,
-          );
-        }
+      forwarded(txHash, success) {
```
Comment (turadg) [inline packages/fast-usdc/src/exos/status-manager.js:385]: boolean params are a smell IME.

I think this function isn't worth its weight anymore. just inline the `publishTxnRecord` call

File: packages/fast-usdc/src/exos/advancer.js (hunk 17 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/advancer.js]: This was tricky to name since "checkForwarded" does more than checking - it actually starts a Forward via MsgTransfer if conditions are met. That's why i wanted to include a verb.

I've heeded your suggestion, and renamed to `checkIfMinted`. (the forward hasn't occurred yet, and the early mint is the thing we're checking)

File: packages/fast-usdc/src/exos/advancer.js (hunk 19 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/advancer.js]: Nice, ty

File: packages/fast-usdc/src/exos/settler.js (hunk 20 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/settler.js:85]: This actually only gets called once if we pull in the vstorage commit. I am going to pull that change into this PR and leave this as is.

File: packages/fast-usdc/src/exos/settler.js (hunk 4 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/settler.js]: It'd take some time to think of an attack vector - currently coming up short.

In general, they're pretty distinct behaviors / states so commingling didn't seem right. 

Since we don't currently have sufficient motivation to separate, I'll leave as-is for now and remove the comment.

File: packages/fast-usdc/src/exos/settler.js (hunk 10 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/settler.js]: Closing 10750 and collapsing into this PR

File: packages/fast-usdc/src/exos/status-manager.js (hunk 22 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/status-manager.js:385]: > just inline the publishTxnRecord call

I don't follow this suggestion - can you rephrase?

File: packages/fast-usdc/src/exos/status-manager.js (hunk 12 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/status-manager.js]: How's this:

```js
/**
 * Remove and return the oldest pending settlement transaction that matches the given
 * forwarding account and amount. Since multiple pending transactions may exist with
 * identical (account, amount) pairs, we process them in FIFO order.
 */
```

File: packages/fast-usdc/src/exos/status-manager.js [hunk 23]
```diff
@@ -247,6 +244,43 @@ export const prepareStatusManager = (
...
+       * If minted while advancing, publish a status update for the advance
+       * to vstorage.
+       *
+       * Does not add or amend `pendingSettleTxs` as this has
+       * already settled.
+       *
+       * @param {EvmHash} txHash
+       * @param {boolean} success whether the Transfer succeeded
+       */
+      advanceOutcomeForSettled(txHash, success) {
```
Comment (0xpatrickdev) [inline packages/fast-usdc/src/exos/status-manager.js]: Not crazy about `advanceOutcomeForSettled`, but implemented [the suggestion](https://github.com/Agoric/agoric-sdk/pull/10750#discussion_r1894133551).

I liked `notify` since all we're doing here is a vstorage write. Technically, it's not settled yet (we could say minted).

Where does that leave us with `notifyObserved`?

File: packages/fast-usdc/src/exos/advancer.js (hunk 17 above)
Comment (turadg) [reply packages/fast-usdc/src/exos/advancer.js]: > it actually starts a Forward via MsgTransfer if conditions are met. That's why i wanted to include a verb.

It happens to but needn't be a commitment to the caller. The caller is providing information and asking for a state.

File: packages/fast-usdc/src/exos/status-manager.js (hunk 22 above)
Comment (turadg) [reply packages/fast-usdc/src/exos/status-manager.js:385]: remove this `forwarded` function and in the places it's called do what it was doing.

File: packages/fast-usdc/src/exos/settler.js (hunk 21 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/settler.js]: TODO will appear in commit history, but not the final HEAD of this PR

File: packages/fast-usdc/src/exos/status-manager.js (hunk 23 above)
Comment (turadg) [reply packages/fast-usdc/src/exos/status-manager.js]: >I liked notify since all we're doing here is a vstorage write.

That shouldn't be a commitment to the caller. The caller tells the status manager about stuff and the status manager abstracts what it does with that information.

> Technically, it's not settled yet (we could say minted).

Yeah that would be better.

>Where does that leave us with notifyObserved?

To be renamed for the same reasons

File: packages/fast-usdc/src/exos/status-manager.js (hunk 22 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/status-manager.js:385]: Ok, but the Settler does not have access to `StatusManager`'s vstorage writer currently. Are you suggesting we export `publishTxnRecord`?

Comment (turadg) [reply packages/fast-usdc/src/exos/status-manager.js:385]: I missed that this was an external method. my mistake! resolving

File: packages/fast-usdc/src/exos/settler.js (hunk 15 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/settler.js]: done

File: packages/fast-usdc/src/exos/status-manager.js (hunk 23 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/status-manager.js]: I landed on `advanceOutcomeForMintedEarly` and `advanceOutcomeForUnknownMint`. If we want something more concise, i think we can tackle that in https://github.com/Agoric/agoric-sdk/issues/10432

File: packages/fast-usdc/src/exos/status-manager.js [hunk 24]
```diff
@@ -303,34 +334,35 @@ export const prepareStatusManager = (
...
+        // extract first item
+        const { status, txHash } = pending[0];
 
+        // Update the pendingSettleTxs map
         if (pending.length > 1) {
           const pendingCopy = [...pending];
-          pendingCopy.splice(dequeueIdx, 1);
-          pendingTxs.set(key, harden(pendingCopy));
+          pendingCopy.shift(); // Remove first element
+          pendingSettleTxs.set(key, harden(pendingCopy));
```
Comment (turadg) [inline packages/fast-usdc/src/exos/status-manager.js]: we shouldn't need to mutate to select elements from an array.
```suggestion
          const [_, ...remaining] = pending;
          pendingSettleTxs.set(key, harden(remaining));
```

**Review by turadg — APPROVED** [review]

File: packages/fast-usdc/src/exos/status-manager.js (hunk 24 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/status-manager.js]: Agree this is awkward, but more so mutating when already we already made a copy. Updated.
